### PR TITLE
[Pending] TW-1148:clicking x while uploading an image selects the message

### PR DESCRIPTION
**Pull request by Te-Z**
Issue: https://github.com/linagora/twake-on-matrix/issues/1148

Resolution:
- add cancel token to cancel the upload
- remove event after cancellation 

[Capture vidéo du 06-03-2024 18:21:05.webm](https://github.com/linagora/twake-on-matrix/assets/31937920/82ab1bf0-a46a-4d31-812e-ff3695a5fcbc)

### CODE REVIEW

File: lib/pages/chat/events/sending_image_info_widget.dart [hunk 1]
```diff
@@ -119,4 +128,9 @@ class SendingImageInfoWidget extends StatelessWidget {
       ),
     );
   }
+
+  void _cancelSending() {
+    mediaCancelToken?.cancel();
+    event.remove();
```
Comment (sherlockvn) [inline lib/pages/chat/events/sending_image_info_widget.dart:134]: the return type should be Future<void>, what if event.remove() failed, i see it have return type Future<bool>, you should check the condition and handle error as well

File: lib/pages/chat/chat.dart [hunk 2]
```diff
@@ -1279,6 +1280,8 @@ class ChatController extends State<Chat>
     }
   }
 
+  final CancelToken mediaCancelToken = CancelToken();
```
Comment (sherlockvn) [inline lib/pages/chat/chat.dart:1283]: i think it may not work well in case of multiple images get cancel, try send multiple images

Comment (hoangdat) [reply lib/pages/chat/chat.dart:1283]: we should have the better way for the case of multiple uploading

**Comment by Te-Z** [comment]
@hoangdat @sherlockvn multiple uploading now handled


https://github.com/linagora/twake-on-matrix/assets/31937920/2133e49f-d4e4-4458-a76f-2ada9bff9817

**Comment by Te-Z** [comment]
> * create an ADR for your modification

There you go @hoangdat . Please let me know if I have to change something on it since it's the first time a write one.

File: pubspec.lock
Comment (hoangdat) [inline pubspec.lock:1]: a lot of unnecessary change for it. please revert

File: assets/l10n/intl_en.arb [hunk 3]
```diff
@@ -3008,5 +3008,6 @@
   "unselect": "Unselect",
   "searchContacts": "Search contacts",
   "tapToAllowAccessToYourMicrophone": "You can enable microphone access in the Settings app to make voice in",
-  "showInChat": "Show in chat"
+  "showInChat": "Show in chat",
+  "cannotCancelImageUpload": "Cannot cancel image update"
```
Comment (hoangdat) [inline assets/l10n/intl_en.arb:3012]: why `update`? 
```suggestion
  "cannotCancelImageUpload": "Upload Cancellation Unsuccessful"
```

File: lib/domain/usecase/send_images_interactor.dart [hunk 4]
```diff
@@ -25,17 +27,20 @@ class SendMediaInteractor {
...
         }
 
-        await room.sendFileEvent(
+        room.sendFileEvent(
           fakeSendingFileInfo.fileInfo,
           msgType: fakeSendingFileInfo.messageType,
           fakeImageEvent: fakeSendingFileInfo.fakeImageEvent,
           shrinkImageMaxDimension: 1600,
           txid: txId,
+          cancelToken: cancelTokens[fileIndex],
```
Comment (hoangdat) [inline lib/domain/usecase/send_images_interactor.dart:43]: what happen if `cancelTokens` is empty?

File: lib/pages/chat/chat_event_list.dart [hunk 5]
```diff
@@ -145,6 +146,9 @@ class ChatEventList extends StatelessWidget {
                   final nextEvent = index < controller.timeline!.events.length
                       ? controller.timeline!.events[currentEventIndex + 1]
                       : null;
+
+                  _cleanCancelTokenMap(event);
+
```
Comment (hoangdat) [inline lib/pages/chat/chat_event_list.dart:151]: I still not get your point? why put it here the cleanup?

File: lib/pages/chat/chat_event_list.dart [hunk 6]
```diff
@@ -203,6 +211,41 @@ class ChatEventList extends StatelessWidget {
...
   }
 
+  void _cleanCancelTokenMap(Event event) {
+    if (event.status == EventStatus.sent &&
+        controller.mediaCancelTokenMapNotifier.value[event.body] != null) {
+      controller.mediaCancelTokenMapNotifier.value.remove(event.body);
+    }
+  }
+
+  Future<void> _cancelSending({
```
Comment (hoangdat) [inline lib/pages/chat/chat_event_list.dart:221]: it is a stateless widget, why we should keep logic here?

File: lib/pages/chat/chat.dart [hunk 7]
```diff
@@ -1294,20 +1298,37 @@ class ChatController extends State<Chat>
...
+      onCameraPicked: (_) => _onSendTap(imagePickerController),
       captionController: _captionsController,
       focusSuggestionController: _focusSuggestionController,
       typeAheadKey: _chatMediaPickerTypeAheadKey,
     );
   }
 
+  void _onSendTap(ImagePickerGridController imagePickerController) {
+    mediaCancelTokenMapNotifier.value = {};
+    final tempMap = <String, CancelToken>{};
```
Comment (sherlockvn) [inline lib/pages/chat/chat.dart:1313]: In this case, the old cancelToken can be removed if the user sends photos twice in a row

File: lib/pages/chat/chat.dart [hunk 8]
```diff
@@ -1294,20 +1298,37 @@ class ChatController extends State<Chat>
...
+
+    for (final selection in imagePickerController.selectedAssets) {
+      final fileName = selection.asset.title;
+      if (fileName != null) {
+        tempMap[fileName] = CancelToken();
```
Comment (sherlockvn) [inline lib/pages/chat/chat.dart:1318]: the key should be `{eventId}_{fileName}`, because user may send the same image 2 times

File: lib/pages/chat/chat.dart [hunk 9]
```diff
@@ -174,6 +175,9 @@ class ChatController extends State<Chat>
 
   final ValueNotifier<DateTime?> stickyTimestampNotifier = ValueNotifier(null);
 
+  final ValueNotifier<Map<String, CancelToken>> mediaCancelTokenMapNotifier =
+      ValueNotifier({});
```
Comment (sherlockvn) [inline lib/pages/chat/chat.dart:179]: i worried that if user back to chat list, this map becomes to empty again

File: assets/l10n/intl_en.arb (hunk 3 above)
Comment (Te-Z) [reply assets/l10n/intl_en.arb:3012]: I was wrong, I wanted to write `upload`

File: pubspec.lock
Comment (Te-Z) [reply pubspec.lock:1]: Those changes are automatic and due to semantic versioning of some packages inside pubspec.yaml. I can revert but those changes will happen anyway when deploying.
Maybe we should add `pubspec.lock` in `.gitignore` since it is a generated file created each time there is a `flutter pub get` ?

File: lib/pages/chat/chat_event_list.dart (hunk 6 above)
Comment (Te-Z) [reply lib/pages/chat/chat_event_list.dart:221]: It is here because:
- it has a strong relationship with the Message widget, so I don't think we should use a callback that needs to propagate back to the controller. I'm afraid it might reduce the code's readability.
- it needs the event used in the Message widget
- it does not need a `setState()` call, so for me no need to replace it with a `StatefulWidget`

WDYT ?

File: lib/pages/chat/chat.dart (hunk 9 above)
Comment (Te-Z) [reply lib/pages/chat/chat.dart:179]: If the user go back to chat list while some uploadings are not done, should those uploads continue or not @hoangdat ?

File: lib/pages/chat/chat_event_list.dart (hunk 5 above)
Comment (Te-Z) [reply lib/pages/chat/chat_event_list.dart]: clean up moved to `chat.dart`

File: lib/pages/chat/chat.dart (hunk 7 above)
Comment (Te-Z) [reply lib/pages/chat/chat.dart]: fixed

File: lib/pages/chat/chat.dart (hunk 8 above)
Comment (Te-Z) [reply lib/pages/chat/chat.dart]: Smart idea. I implemented it

File: lib/pages/chat/chat.dart (hunk 9 above)
Comment (sherlockvn) [reply lib/pages/chat/chat.dart:179]: The uploading is still continue, because it's registered under getIt as a singleton.

Comment (sherlockvn) [reply lib/pages/chat/chat.dart:179]: In order to to fix this case, i think you need to have something like a download manager, register it under the getIt, do all the stuff releated to download here

File: pubspec.yaml [hunk 10]
```diff
@@ -232,7 +232,7 @@ msix_config:
 
 dependency_overrides:
   # Until all dependencies are compatible. Missing: file_picker_cross, flutter_matrix_html
-  ffi: ^2.0.0
+  ffi: 2.1.0
```
Comment (hoangdat) [inline pubspec.yaml:235]: why no change for local_notification?

File: pubspec.lock
Comment (hoangdat) [inline pubspec.lock:1]: why too much change here?
